src/tiled/objecttemplatemodel.cpp

```diff
@@ -96,17 +103,53 @@ QVariant ObjectTemplateModel::data(const QModelIndex &index, int role) const
 
 bool ObjectTemplateModel::addNewDocument(QString fileName, QString *error)
 {
-    // TODO: remove the old file if it was already loaded
+    // Remove old document if the new document overwrites it
+    QMutableListIterator<TemplateGroupDocument*> it(mTemplateDocuments);
+    while (it.hasNext()) {
+        if (it.next()->fileName() == fileName) {
+            int pos = mTemplateDocuments.indexOf(it.value());
```

Comment:
Instead of iterating the list forwards and for each match, finding the index, you could instead iterate the list backwards like `for (int i = mTemplateDocuments.size() - 1; i >= 0; --i) {`.